File: LeetCodePython/PalindromeNumber.py
"""
Given an integer x, return true if x is palindrome integer.

An integer is a palindrome when it reads the same backward as forward.

For example, 121 is a palindrome while 123 is not.

Example 1:
Input: x = 121
Output: true
Explanation: 121 reads as 121 from left to right and from right to left.

Example 2:
Input: x = -121
Output: false
Explanation: From left to right, it reads -121. From right to left, it becomes 121-. Therefore it is not a palindrome.
Example 3:

Input: x = 10
Output: false
Explanation: Reads 01 from right to left. Therefore it is not a palindrome.


Constraints:
-231 <= x <= 231 - 1
Follow up: Could you solve it without converting the integer to a string?

hint:
1 Beware of overflow when you reverse the integer.
"""
class PalindromeNumber:
    def isPalindrome(self, x: int) -> bool:
        if x < 0 or (x > 0 and x % 10 == 0):
            return False
        rev = 0
        while x > rev:
            rev = rev * 10 + x % 10
            x = x // 10
        return True if (x == rev or x == rev // 10) else False

File: LeetCodePractice/src/Math/PalindromeNumber.java
package Math;

public class PalindromeNumber {
	public boolean isPalindrome(int x) {
        if (x < 0) {
			return false;
		}
		int highest = 1;
		while (x / highest >= 10) {
			highest *= 10;
		}

		while (x != 0) {
			int right = x % 10;
			int left = x / highest;
			if (right == left) {
				x = x % highest;
				x = x / 10;
				highest = highest / 100;
			} else {
				return false;
			}
		}
		return true;
    }
}
